Add a field to Delayed for storing a clean up action

alpmestan opened the pull request:
I need something like this in [servant-multipart](https://github.com/haskell-servant/servant-multipart) in order to delete files created in a temporary dir once the handler has run.

This is not the definitive code, but at least we can start a discussion.

For instance, @jkarni pointed out that the clean up might not always run with the code as is because of exceptions.

I decided to make the clean up actions live in `IO` (and not `DelayedIO`) because most of the time, if not always, the code we want to run there lives in `IO` and doesn't really need access to the request, but rather to something created in the body.

jkarni commented on servant-server/src/Servant/Server/Internal/RoutingApplication.hs:
```diff
@@ -258,8 +271,11 @@ runAction :: Delayed env (Handler a)
+runAction action env req respond k = do
+  (routeResult, cleanup) <- runDelayed action env req
+  resp <- go routeResult
+  cleanup
```
I think this should be `bracket (runDelayed action env eq) snd (go . fst)`?

jkarni commented on servant-server/src/Servant/Server/Internal/RoutingApplication.hs:
```diff
@@ -103,6 +105,10 @@ data Delayed env c where
              , authD     :: DelayedIO auth
              , bodyD     :: DelayedIO body
              , serverD   :: captures -> auth -> body -> Request -> RouteResult c
+             , cleanupD  :: body -> IO ()
```
As I mentioned elsewhere, I don't know exactly why you need the `body` arg. Or why *just* the `body`, for that matter. `DelayedIO (IO ())` seems like a more intuitive signature, to me.

alpmestan replied:
1/ Why is it more intuitive? Clean up will rarely (if ever) need to look at the raw request.

2/ In my case, I get filepaths of temp files and what not by looking at `body`. We can indeed add other arguments to cleanupD, but not make it a simple action, because I otherwise don't have any mean for my cleanup thing to get its hands on the temp files it is supposed to delete. More generally, this feature is useful anytime we want to clean up after body checks and other stuffs (anything, really). Except maybe for somewhat "global" stuffs, we need a way to for the clean up action to get its hands on whatever it is that needs cleaning up, so I use an argument for that.

I would not mind adding more arguments (results of the other checks). But making it a `DelayedIO` so that I have to parse everything from the request again? No.

alpmestan commented:
Yes, I'll write tests before it's approved for merging.

soenkehahn commented on servant-server/src/Servant/Server/Internal/RoutingApplication.hs:
```diff
@@ -196,8 +203,9 @@ addBodyCheck :: Delayed env (a -> b)
-    , serverD = \ c a (z, v) req -> ($ v) <$> serverD c a z req
+    { bodyD    = (,) <$> bodyD <*> new
+    , serverD  = \ c a (z, v) req -> ($ v) <$> serverD c a z req
+    , cleanupD = cleanupD . fst -- not sure that's right
```
Seems right to me. :)

soenkehahn commented on servant-server/src/Servant/Server/Internal/RoutingApplication.hs:
```diff
@@ -240,13 +248,19 @@ passToServer Delayed{..} x =
+        methodD
+        a <- authD
+        b <- bodyD
+        liftIO (writeIORef cleanupRef $ cleanupD b)
```
I would feel better if we were able to use something like `bracket` for this. Not sure that's easily doable.

alpmestan replied:
For writing the `IORef`, or for the whole sequence of operations?

alpmestan replied:
Note that this is going to change a bit. @jkarni pointed out that adding clean up actions can happen from within `bodyD` and that we should not wait for it to terminate in order to add a clean up action, because if an exception occurs the cleanup never gets run. Particularly annoying in the case of file uploads where files can pile up in /tmp.

alpmestan commented:
@soenkehahn Yes, I'll add tests that mimic my use case as far as Delayed/cleanup are concerned. In the meantime, you might want to take a look here: https://github.com/haskell-servant/servant-multipart/blob/master/src/Servant/Multipart.hs#L267

jkarni commented on servant-server/src/Servant/Server/Internal/RoutingApplication.hs:
```diff
@@ -112,12 +115,33 @@ instance Functor (Delayed env) where
+newCleanupRef = CleanupRef <$> newIORef (return ())
+
+-- | Add a clean up action to a 'CleanupRef'
+addCleanup' :: IO () -> CleanupRef -> IO ()
```
I'd just have this is a `where` clause to `addCleanup` so it's obvious that it's not meant to be used elsewhere.

phadej replied:
I agree

jkarni commented on servant-server/src/Servant/Server/Internal.hs:
```diff
@@ -398,11 +399,17 @@ instance HasServer Raw context where
-      Route app   -> app request (respond . Route)
-      Fail a      -> respond $ Fail a
-      FailFatal e -> respond $ FailFatal e
+    -- note: a Raw application doesn't register any cleanup
```
I *think* this is not just for consistency. E.g.:

```
type API = FileUpload :> Raw
```

Isn't in this case not having this going to cause the cleanup action to not be run? (It probably makes sense to have a test for (1) cleanup being run for `Raw`; (2) cleanup only being run *once*)

alpmestan commented:
Added some basic tests to make sure that clean up operations are run under various circumstances, among which the presence of exceptions in `bodyD` or `serverD`. Also rebased against master. Could anyone take a look and let me know what's missing?

phadej commented on servant-server/test/Servant/Server/Internal/RoutingApplicationSpec.hs:
```diff
@@ -0,0 +1,51 @@
+  , methodD = DelayedIO $ \_req_ _cl -> ok
+  , authD = DelayedIO $ \_req _cl -> ok
+  , bodyD = do
+      liftIO (writeFile "delayed.test" "hia")
```
I'd prefer `IORef`, so the tests don't write to disk.

jkarni commented:
@phadej @alpmestan is this superseded by more recent PRs? Can we close?

phadej commented:
AFAICS all comments here are addressed already.
